### PR TITLE
Improve scheduling of new missions during return to home

### PR DESCRIPTION
**Pull request by mrica-equinor**

### CODE REVIEW

File: backend/api/EventHandlers/MissionEventHandler.cs [hunk 1]
```diff
@@ -347,18 +347,14 @@ public async Task CancelReturnToHomeOnNewMissionSchedule(MissionRun missionRun)
 
                 if (returnToHomeMission.Status != MissionStatus.Pending)
```
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:348]: Don't we want to abort pending mission runs as well?

File: backend/api/EventHandlers/MqttEventHandler.cs [hunk 2]
```diff
@@ -300,14 +300,18 @@ private async void OnIsarMissionUpdate(object? sender, MqttReceivedArgs mqttArgs
 
             if (flotillaMissionRun.Status == status) { return; }
 
-            MissionRun updatedFlotillaMissionRun;
-            try { updatedFlotillaMissionRun = await MissionRunService.UpdateMissionRunStatusByIsarMissionId(isarMission.MissionId, status); }
-            catch (MissionRunNotFoundException) { return; }
+            var updatedFlotillaMissionRun = flotillaMissionRun;
```
Comment (oysand) [inline backend/api/EventHandlers/MqttEventHandler.cs:303]: Why assign it here? It is overwritten again at line 307?

File: backend/api/EventHandlers/MissionEventHandler.cs [hunk 3]
```diff
@@ -347,18 +347,14 @@ public async Task CancelReturnToHomeOnNewMissionSchedule(MissionRun missionRun)
...
                 {
-                    try { await MissionScheduling.StopCurrentMissionRun(missionRun.Robot.Id); }
+                    try
+                    {
+                        await MissionService.UpdateMissionRunProperty(returnToHomeMission.Id, "Status", MissionStatus.Aborted);
```
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:352]: Move above if pending

File: backend/api/EventHandlers/MissionEventHandler.cs (hunk 1 above)
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:348]: change to a guard statement

File: backend/api/EventHandlers/MissionEventHandler.cs [hunk 4]
```diff
@@ -347,18 +347,14 @@ public async Task CancelReturnToHomeOnNewMissionSchedule(MissionRun missionRun)
 
                 if (returnToHomeMission.Status != MissionStatus.Pending)
                 {
-                    try { await MissionScheduling.StopCurrentMissionRun(missionRun.Robot.Id); }
+                    try
```
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:350]: do the try block after the guard statement

File: backend/api.test/EventHandlers/TestMissionEventHandler.cs [hunk 5]
```diff
@@ -526,8 +526,7 @@ public async Task ReturnHomeMissionCancelledIfNewMissionScheduled()
 
             // Assert
             var updatedReturnHomeMission = await _missionRunService.ReadById(returnToHomeMission.Id);
-            Assert.True(updatedReturnHomeMission?.Status.Equals(MissionStatus.Cancelled));
-            Assert.True(updatedReturnHomeMission?.Tasks.FirstOrDefault()?.Status.Equals(Api.Database.Models.TaskStatus.Cancelled));
+            Assert.True(updatedReturnHomeMission?.Status.Equals(MissionStatus.Aborted));
```
Comment (oysand) [inline backend/api.test/EventHandlers/TestMissionEventHandler.cs:529]: Rename test

File: backend/api/Services/ReturnToHomeService.cs [hunk 6]
```diff
@@ -111,5 +111,49 @@ private async Task<MissionRun> ScheduleReturnToHomeMissionRun(string robotId)
...
+            IList<MissionStatus> missionStatuses = [MissionStatus.Ongoing, MissionStatus.Pending, MissionStatus.Paused];
+            var existingReturnToHomeMissions = await missionRunService.ReadMissionRuns(missionRun.Robot.Id, MissionRunType.ReturnHome, missionStatuses);
+
+            if (existingReturnToHomeMissions.Count == 1 && existingReturnToHomeMissions[0].Id != missionRun.Id)
+            {
+                var returnToHomeMission = existingReturnToHomeMissions[0];
+
+                try
+                {
+                    if (!await localizationService.RobotIsOnSameDeckAsMission(missionRun.Robot.Id, missionRun.Area.Id))
```
Comment (oysand) [inline backend/api/Services/ReturnToHomeService.cs:126]: This is checked in the controller, can be removed here

Comment (oysand) [inline backend/api/Services/ReturnToHomeService.cs:126]: Add one of these checks to the create endpoint for custom mission

File: backend/api.test/EventHandlers/TestMissionEventHandler.cs [hunk 7]
```diff
@@ -505,7 +505,7 @@ public async Task ReturnHomeMissionNotScheduledIfRobotIsNotLocalized()
         }
 
         [Fact]
-        public async Task ReturnHomeMissionCancelledIfNewMissionScheduled()
+        public async Task ReturnHomeMissionAbortedIfNewMissionScheduled()
```
[the remark this answers is not in the document]
Comment (mrica-equinor) [reply backend/api.test/EventHandlers/TestMissionEventHandler.cs:508]: There was a test already i just had to adjust it to the new code

File: backend/api/EventHandlers/MissionEventHandler.cs [hunk 8]
```diff
@@ -86,9 +89,11 @@ private async void OnMissionRunCreated(object? sender, MissionRunCreatedEventArg
             }
             _scheduleLocalizationSemaphore.Release();
 
-            await CancelReturnToHomeOnNewMissionSchedule(missionRun);
-
             _startMissionSemaphore.WaitOne();
+
+            var activeReturnToHomeMission = await ReturnToHomeService.GetActiveReturnToHomeOnSameDeckAsMission(missionRun);
```
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:94]: Only pass robot id

File: backend/api/EventHandlers/MissionEventHandler.cs [hunk 9]
```diff
@@ -86,9 +89,11 @@ private async void OnMissionRunCreated(object? sender, MissionRunCreatedEventArg
...
+            if (activeReturnToHomeMission != null) { await MissionScheduling.AbortReturnToHomeMission(activeReturnToHomeMission); }
```
Comment (oysand) [inline backend/api/EventHandlers/MissionEventHandler.cs:95]: Only pass robot id

**Review by oysand — APPROVED** [review]
LGTM

**Pull request merged**